install-from-github.js

```diff
@@ -24,41 +24,69 @@ try {
   });
 } catch (e) {
 }
-const {downloadBrowser} = require('./download-browser');
-const playwright = require('.');
+
+const path = require('path');
+const fs = require('fs');
+const util = require('util');
+const rmAsync = util.promisify(require('rimraf'));
+const existsAsync = path => fs.promises.access(path).then(() => true, e => false);
+const {downloadBrowserWithProgressBar} = require('./download-browser');
+const protocolGenerator = require('./utils/protocol-types-generator');
+const packageJSON = require('./package.json');
+
+const DOWNLOADED_BROWSERS_JSON_PATH = path.join(__dirname, '.downloaded-browsers.json');
+const DOWNLOAD_PATHS = {
+  chromium: path.join(__dirname, '.local-browsers', `chromium-${packageJSON.playwright.chromium_revision}`),
+  firefox: path.join(__dirname, '.local-browsers', `firefox-${packageJSON.playwright.firefox_revision}`),
+  webkit: path.join(__dirname, '.local-browsers', `webkit-${packageJSON.playwright.webkit_revision}`),
+};
 
 (async function() {
-  const protocolGenerator = require('./utils/protocol-types-generator');
+  const downloadedBrowsersJSON = await fs.promises.readFile(DOWNLOADED_BROWSERS_JSON_PATH, 'utf8').then(json => JSON.parse(json)).catch(() => ({}));
   try {
-    const chromeRevision = await downloadAndCleanup(playwright.chromium);
-    await protocolGenerator.generateChromiunProtocol(chromeRevision);
+    if (!(await existsAsync(DOWNLOAD_PATHS.chromium))) {
+      const crExecutablePath = await downloadBrowserWithProgressBar(DOWNLOAD_PATHS.chromium, 'chromium');
+      downloadedBrowsersJSON.crExecutablePath = crExecutablePath;
+      await fs.promises.writeFile(DOWNLOADED_BROWSERS_JSON_PATH, JSON.stringify(downloadedBrowsersJSON));
```

Comment:
Write it after generating the protocol in case it throws?

Reply:
Done.